File: 2. Structural Design Patterns/Proxy/Java/ProxyTemplate.java
// Proxy Pattern – Java Template

interface Subject { void request(); }
class RealSubject implements Subject { public void request(){ /* heavy work */ } }
class Proxy implements Subject { private RealSubject real; public void request(){ if (real == null) real = new RealSubject(); /* checks/logging */ real.request(); } }
public class ProxyTemplate { public static void main(String[] args){ Subject s = new Proxy(); s.request(); } }

File: 2. Structural Design Patterns/Proxy/Kotlin/ProxyTemplate.kt
// Proxy Pattern – Kotlin Template

interface Subject { fun request() }
class RealSubject : Subject { override fun request() { /* heavy work */ } }
class Proxy : Subject { private var real: RealSubject? = null; override fun request(){ if (real == null) real = RealSubject(); /* checks/logging */ real!!.request() } }
fun main(){ val s: Subject = Proxy(); s.request() }
